### PR TITLE
Support `sni_hostname` extension with SOCKS proxy.

### LINKED ISSUE #772: Support 'sni_hostname' extension with SOCKS proxy.
- [x] Initially raised as discussion #771

Currently, there is an issue with `sni_hostname` not functioning correctly when the connection involves a socks proxy.
This problem arises due to the oversight in commit #696, where the update to `_sync/socks_proxy.py` and `_async/socks_proxy.py` was omitted.

## Code for reproduction (written by @karosis88 at #771)
Set up proxy server (defaults to 1080 port)
```bash
git clone https://github.com/itsjfx/python-socks5-server
python python-socks5-server/server.py
```
Client
```python
import httpcore

pool = httpcore.SOCKSProxy("socks5://127.0.0.1:1080")
response = pool.request("GET", 
                        "https://185.199.108.153:443", 
                        headers=[(b'Host', 'www.encode.io')], 
                        extensions={'sni_hostname': 'www.encode.io'})
print(response)
```
Expected output
```python
<Response [200]>
```
Actual output
```python
httpcore.ConnectError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: IP address mismatch, certificate is not valid for '185.199.108.153'. (_ssl.c:1123)
```

## Suggested fix
Update the `_sync/socks_proxy.py` and `_async/socks_proxy.py` in a manner similar to that of the `_sync/connection.py` and `_async/connection.py`.

Example fix for `_async/socks_proxy.py`:
```python
218    218      timeouts = request.extensions.get("timeout", {})
       219   +  sni_hostname = request.extensions.get("sni_hostname", None)
219    220      timeout = timeouts.get("connect", None)

261          -  "server_hostname": self._remote_origin.host.decode("ascii"),
       262   +  "server_hostname": sni_hostname
       263   +  or self._origin.host.decode("ascii"),
```

### PR DESCRIPTION
**Pull request by Allgot**
# Summary

The introduction of a new extension, `sni_hostname`, allows users to define the hostname to be used during the TLS handshake. However, the update was applied inconsistently, leading to an incomplete extension. Currently, the `sni_hostname` extension is not functioning when a SOCKS proxy is employed. This PR aims to address this gap and rectify the issue.

# Checklist

- [x] I understand that this PR may be closed in case there was no previous discussion. (This doesn't apply to typos!)
- [x] I've added a test for each change that was introduced, and I tried as much as possible to make a single atomic change.
- [x] I've updated the documentation accordingly.

* Previous issue: #773

Closes #772

### CODE REVIEW

**Comment by karosis88** [comment]
Great, Thanks. Let's also update the changelog.

File: tests/_async/test_socks_proxy.py [hunk 1]
```diff
@@ -191,3 +191,66 @@ async def test_socks5_request_incorrect_auth():
...
+    ) as proxy:
+        # Sending an intial request, which once complete will return to the pool, IDLE.
+        async with proxy.stream(
+            "GET",
+            "https://93.184.216.34/",
+            headers=[(b"Host", "example.com")],
+            extensions={"sni_hostname": "example.com"},
+        ) as response:
+            info = [repr(c) for c in proxy.connections]
+            assert info == [
```
Comment (karosis88) [inline tests/_async/test_socks_proxy.py:229]: Let's remove this check

File: tests/_async/test_socks_proxy.py [hunk 2]
```diff
@@ -191,3 +191,66 @@ async def test_socks5_request_incorrect_auth():
...
+                "<AsyncSocks5Connection ['https://93.184.216.34:443', HTTP/1.1, ACTIVE, Request Count: 1]>"
+            ]
+            await response.aread()
+
+        assert response.status == 200
+        assert response.content == b"Hello, world!"
+        info = [repr(c) for c in proxy.connections]
+        assert info == [
```
Comment (karosis88) [inline tests/_async/test_socks_proxy.py:237]: And this

**Review by karosis88 — COMMENTED** [review]
Unit tests should only test one specific thing and should ideally have only one assertion. I believe this test is overly complicated, so let us simplify it.

**Comment by karosis88** [comment]
Also, I'm not sure if this test actually checks whether sni_hostname works or not, so let's just remove it entirely.

File: tests/_async/test_socks_proxy.py [hunk 3]
```diff
@@ -191,3 +191,66 @@ async def test_socks5_request_incorrect_auth():
...
+        )
+        assert not proxy.connections[0].can_handle_request(
+            httpcore.Origin(b"http", b"other.com", 80)
+        )
+        assert proxy.connections[0].can_handle_request(
+            httpcore.Origin(b"https", b"93.184.216.34", 443)
+        )
+        assert not proxy.connections[0].can_handle_request(
+            httpcore.Origin(b"https", b"other.com", 443)
+        )
```
Comment (tomchristie) [inline tests/_async/test_socks_proxy.py:256]: These tests, don't. So...

```suggestion
```

File: tests/_sync/test_socks_proxy.py [hunk 4]
```diff
@@ -191,3 +191,66 @@ def test_socks5_request_incorrect_auth():
...
+        )
+        assert not proxy.connections[0].can_handle_request(
+            httpcore.Origin(b"http", b"other.com", 80)
+        )
+        assert proxy.connections[0].can_handle_request(
+            httpcore.Origin(b"https", b"93.184.216.34", 443)
+        )
+        assert not proxy.connections[0].can_handle_request(
+            httpcore.Origin(b"https", b"other.com", 443)
+        )
```
Comment (tomchristie) [inline tests/_sync/test_socks_proxy.py:256]:
```suggestion
```

**Review by tomchristie — APPROVED** [review]

**Review by karosis88 — APPROVED** [review]

**Comment by tomchristie** [comment]
Thanks @Allgot.

**Pull request merged**